### PR TITLE
add s3 select api

### PR DESCRIPTION
**Pull request by balamurugana**
Below new method is added
```java
public SelectResponseStream selectObjectContent(String bucketName, String objectName, String sqlExpression,
                                        InputSerialization is, OutputSerialization os, boolean requestProgress,
                                        Long scanStartRange, Long scanEndRange, ServerSideEncryption sse)
```

### CODE REVIEW

File: api/src/main/java/io/minio/messages/SelectObjectContentRequest.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+
+  /**
+   * Constructs new SelectObjectContentRequest object for given parameters.
+   */
+  public SelectObjectContentRequest(String expression, boolean requestProgress, InputSerialization is,
+                                    OutputSerialization os, Long scanStartRange, Long scanEndRange)
+    throws XmlPullParserException {
+    super();
+    this.name = "SelectObjectContentRequest";
```
Comment (sinhaashish) [inline api/src/main/java/io/minio/messages/SelectObjectContentRequest.java]: Where is the name defined in this class ? Ran an example code and its erroring out as `java.lang.IllegalArgumentException`

Comment (balamurugana) [reply api/src/main/java/io/minio/messages/SelectObjectContentRequest.java]: My suggestion is to explore the source code and find how XML encoding/decoding is done in minio-java. If you have any difficulties, let me know.

Please report any error with appropriate traces including your java environment detail with source code.

**Comment by harshavardhana** [comment]
@balamurugana can you fix the checkstyle failures?

```
Download https://repo.maven.apache.org/maven2/antlr/antlr/2.7.7/antlr-2.7.7.jar
[ant:checkstyle] /home/runner/work/minio-java/minio-java/functional/FunctionalTest.java:3266: '+' have incorrect indentation level 8, expected level should be 10.
[ant:checkstyle] /home/runner/work/minio-java/minio-java/functional/FunctionalTest.java:3267: '+' have incorrect indentation level 8, expected level should be 10.
[ant:checkstyle] /home/runner/work/minio-java/minio-java/functional/FunctionalTest.java:3268: '+' have incorrect indentation level 8, expected level should be 10.
[ant:checkstyle] /home/runner/work/minio-java/minio-java/functional/FunctionalTest.java:3269: '+' have incorrect indentation level 8, expected level should be 10.
:api:checkstyleMain FAILED

2 actionable tasks: 2 executed
FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':api:checkstyleMain'.
> Checkstyle rule violations were found. See the report at: file:///home/runner/work/minio-java/minio-java/api/build/reports/checkstyle/main.html

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 36s
##[error]Process completed with exit code 1.
```

File: api/src/main/java/io/minio/messages/SelectObjectContentRequest.java (hunk 1 above)
Comment (sinhaashish) [reply api/src/main/java/io/minio/messages/SelectObjectContentRequest.java]: java version `openjdk version "11.0.5" 2019-10-15`
The below code is from this PR `API.md` file. 
 ``` 
 try {
      /* play.min.io for test and development. */
      MinioClient minioClient = new MinioClient("https://play.min.io", "Q3AM3UQ867SPQQA43P2F", "zuf+tfteSlswRu7BJ86wekitnifILbZam1KYY3TG");
      minioClient.traceOn(System.out);
      String sqlExpression = "select * from S3Object";
      InputSerialization is = InputSerialization.csv(null, false, null, null, FileHeaderInfo.USE, null, null, null);
      OutputSerialization os = OutputSerialization.csv(null, null, null, QuoteFields.ASNEEDED, null);
      SelectResponseStream stream = minioClient.selectObjectContent("sinha", "test.csv", sqlExpression, is, os, true, null, null, null);
      byte[] buf = new byte[512];
      int bytesRead = stream.read(buf, 0, buf.length);
      System.out.println(new String(buf, 0, bytesRead, StandardCharsets.UTF_8));
      Stats stats = stream.stats();
      System.out.println("bytes scanned: " + stats.bytesScanned());
      System.out.println("bytes processed: " + stats.bytesProcessed());
      System.out.println("bytes returned: " + stats.bytesReturned());
      stream.close();
    } catch (MinioException e) {
      System.out.println("Error occurred: " + e);
    }
```

Output 
```
/usr/lib/jvm/java-1.11.0-openjdk-amd64/bin/java -javaagent:/opt/idea-IC-191.6183.87/lib/idea_rt.jar=34865:/opt/idea-IC-191.6183.87/bin -Dfile.encoding=UTF-8 -classpath /home/ashish/code/go/src/github.com/minio/minio-java/api/build/libs/minio-6.0.12-DEV-all.jar:/home/ashish/client/target/classes SelectObjectExample
---------START-HTTP---------
GET /sinha?location= HTTP/1.1
Host: play.min.io
User-Agent: MinIO (amd64; amd64) minio-java/dev
x-amz-content-sha256: UNSIGNED-PAYLOAD
x-amz-date: 20191221T015201Z
Authorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20191221/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=*REDACTED*

HTTP/1.1 200
Server: nginx/1.10.3 (Ubuntu)
Date: Sat, 21 Dec 2019 01:52:03 GMT
Content-Type: application/xml
Content-Length: 137
Connection: keep-alive
Accept-Ranges: bytes
Content-Security-Policy: block-all-mixed-content
Vary: Origin
X-Amz-Bucket-Region: us-east-1
X-Amz-Request-Id: 15E23FF0EAC2F5B8
X-Xss-Protection: 1; mode=block

----------END-HTTP----------
Exception in thread "main" java.lang.IllegalArgumentException
	at com.google.api.client.repackaged.com.google.common.base.Preconditions.checkArgument(Preconditions.java:108)
	at com.google.api.client.util.Preconditions.checkArgument(Preconditions.java:37)
	at com.google.api.client.util.DataMap.<init>(DataMap.java:45)
	at com.google.api.client.util.Data.mapOf(Data.java:186)
	at com.google.api.client.xml.XmlNamespaceDictionary.computeAliases(XmlNamespaceDictionary.java:248)
	at com.google.api.client.xml.XmlNamespaceDictionary.computeAliases(XmlNamespaceDictionary.java:266)
	at com.google.api.client.xml.XmlNamespaceDictionary.computeAliases(XmlNamespaceDictionary.java:266)
	at com.google.api.client.xml.XmlNamespaceDictionary.computeAliases(XmlNamespaceDictionary.java:266)
	at com.google.api.client.xml.XmlNamespaceDictionary.startDoc(XmlNamespaceDictionary.java:236)
	at com.google.api.client.xml.XmlNamespaceDictionary.serialize(XmlNamespaceDictionary.java:227)
	at com.google.api.client.xml.XmlNamespaceDictionary.toStringOf(XmlNamespaceDictionary.java:174)
	at com.google.api.client.xml.GenericXml.toString(GenericXml.java:66)
	at io.minio.MinioClient.executeReq(MinioClient.java:1092)
	at io.minio.MinioClient.execute(MinioClient.java:1078)
	at io.minio.MinioClient.selectObjectContent(MinioClient.java:5818)
	at SelectObjectExample.main(SelectObjectExample.java:48)

```
Can you tell what wrong i am doing here.

Comment (balamurugana) [reply api/src/main/java/io/minio/messages/SelectObjectContentRequest.java]: 1. Java 11 is not supported yet.
2. Please pull latest code from this PR and test with java 8/9/10.

**Comment by kannappanr** [comment]
@balamurugana Functional test is failing

```
Test: selectObjectContent(String bucketName, String objectName, String sqlExpression, InputSerialization is, OutputSerialization os, boolean requestProgress, Long scanStartRange, Long scanEndRange, ServerSideEncryption sse)
java.lang.Exception: result mismatch; expected: 1997,Ford,E350,"ac, abs, moon",3000.00
1999,Chevy,"Venture ""Extended Edition""","",4900.00
1999,Chevy,"Venture ""Extended Edition, Very Large""",,5000.00
1996,Jeep,Grand Cherokee,"MUST SELL!
air, moon roof, loaded",4799.00
, got: 1997,Ford,E350,"ac, abs, moon",3000.00
1999,Chevy,"Venture ""Extended Edition""",,4900.00
1999,Chevy,"Venture ""Extended Edition, Very Large""",,5000.00
1996,Jeep,Grand Cherokee,"MUST SELL!
air, moon roof, loaded",4799.00

	at FunctionalTest.selectObjectContent_test1(FunctionalTest.java:3284)
	at FunctionalTest.runTests(FunctionalTest.java:3395)
	at FunctionalTest.main(FunctionalTest.java:3514)
:functional:runFunctionalTest FAILED
```

**Comment by kannappanr** [comment]
ping @sinhaashish @donatello @nitisht

**Comment by donatello** [comment]
@balamurugana Could you fix the tests so that the build succeeds?

**Comment by balamurugana** [comment]
> @balamurugana Could you fix the tests so that the build succeeds?

I see no problem with tests in this PR. But there is a problem in existing tests introduced in https://github.com/minio/minio-java/pull/727 i.e. if tests listed after `listenBucketNotification_test1()` make hanging due to ever running thread in `listenBucketNotification_test1()` (Just my guess).

**Comment by harshavardhana** [comment]
> I see no problem with tests in this PR. But there is a problem in existing tests introduced in #727 i.e. if tests listed after `listenBucketNotification_test1()` make hanging due to ever running thread in `listenBucketNotification_test1()` (Just my guess).

This in my tests is related to java versions not sure which one @balamurugana

**Comment by kannappanr** [comment]
@donatello @sinhaashish PTAL

**Pull request merged**